### PR TITLE
[#154400504] spid metadata

### PR DESCRIPTION
**Pull request by lussoluca**
The biggest changes are in the [spid-passport](https://github.com/lussoluca/spid-passport) package.

### CODE REVIEW

File: app/src/controllers/authenticationController.js [hunk 1]
```diff
@@ -51,4 +55,24 @@ export default class AuthenticationController {
...
     );
   }
+
+  /**
+   * The metadata for this Service Provider.
+   *
+   * @param req
+   * @param res
+   */
+  metadata(req: express$Request, res: express$Response) {
```
Comment (cloudify) [inline app/src/controllers/authenticationController.js]: accessing global vars makes testing hard, I suggest to refactor this in two steps:

``` js
function getMetadataHandler(certPath, spidStrategy) {
  return (req, res) => {...};
}

// ...

{
  // ...
  metadata: getMetadataHandler(process.env.SAML_CERT_PATH || "./certs/cert.pem", spidStrategy)
}
```

File: app/src/server.js [hunk 2]
```diff
@@ -78,6 +78,10 @@ app.post("/assertionConsumerService", spidAuth, function(
   acsController.acs(req, res);
 });
 
+app.get("/metadata", function(req: express$Request, res: express$Response) {
```
Comment (cloudify) [inline app/src/server.js]:
``` js
app.get("/metadata", acsController.metadata);
```

File: app/src/controllers/authenticationController.js [hunk 3]
```diff
@@ -10,6 +10,20 @@ import spidStrategy from "../strategies/spidStrategy";
 const fs = require("fs");
 const winston = require("winston");
 
+// Retrieves the metadata for this Service Provide from the spidStrategy.
```
Comment (cloudify) [inline app/src/controllers/authenticationController.js]: typo `Provider`

File: app/src/controllers/authenticationController.js [hunk 4]
```diff
@@ -66,13 +80,6 @@ export default class AuthenticationController {
...
 
-    const metadata = spidStrategy.generateServiceProviderMetadata(
-      fs.readFileSync(certPath, "utf-8")
-    );
-
-    res
-      .status(200)
-      .set("Content-Type", "application/xml")
-      .send(metadata);
+    getMetadataHandler(certPath, spidStrategy)(req, res);
```
Comment (cloudify) [inline app/src/controllers/authenticationController.js]: wait sorry I didn't see that this was inside a class, my suggestion to get the actual certificate content and the `spidStrategy` in the controller

reading of the certificate file should happen outside the handler and during server init (i.e. before you instantiate `acsController`)

Comment (lussoluca) [reply app/src/controllers/authenticationController.js]: Ok, I've refactored the file reading

File: app/src/controllers/authenticationController.js [hunk 5]
```diff
@@ -75,11 +58,20 @@ export default class AuthenticationController {
...
+   * @param samlCertFile
+   * @param spidStrategy
    */
-  metadata(req: express$Request, res: express$Response) {
-    const certPath = process.env.SAML_CERT_PATH || "./certs/cert.pem";
-    winston.log("info", "Reading SAML certificate file from %s", certPath);
+  metadata(
+    req: express$Request,
+    res: express$Response,
+    samlCertFile: string,
```
Comment (cloudify) [inline app/src/controllers/authenticationController.js]: can't `samlCertFile` and `spidStrategy` be set in the constructor?

either the life of the `AuthenticationController` class spans multiple calls to `metadata` (and I assume `samlCertFile` and  `spidStrategy`) or `AuthenticationController` and the `metadata` method have the same lifecycle (that you don't need a class bust just a single method)

Comment (cloudify) [reply app/src/controllers/authenticationController.js]: let me know you you want to have a quick call to clarify my point further

File: app/src/server.js [hunk 6]
```diff
@@ -29,12 +29,30 @@ const passport = require("passport");
 
 const port = process.env.PORT || 443;
 
+// Read security files.
```
Comment (cloudify) [inline app/src/server.js]: please add a comment describing the purpose of each file and why they're duplicated

File: app/src/controllers/authenticationController.js (hunk 5 above)
Comment (lussoluca) [reply app/src/controllers/authenticationController.js]: No it's clear, I'll refactor the code a bit to inject all the dependencies directly into the class constructor

**Pull request merged**